### ISSUE TITLE
[Discussion] Sub-namespaces for each standard

### ISSUE THREAD

**Issue by peter-mw** [php-cds/php-cds#26]
Hello,

Inspired by this comment or reddit i decided to bring to your attention the idea of having sub-namespaces for each PCS

https://www.reddit.com/r/PHP/comments/3vu9dy/psr6_is_about_to_pass_and_still_has_serious/cxqs7t7

This will allow the standard to evolve and not being set in stone. 

For example this Cache proposal https://github.com/php-cds/php-cds/issues/7

Its great, but it does not cover the needs of all. For example in my work i need to have the cache separated by tags/groups. 

If we have sub-namespace for the PCS the initial "simplified cache" proposal can go in namespace `\Pcs\Cache` with the name `DefaultCache`, and when i need Tags support i can use an interface named `TaggedCache`

In order to avoid duplication and mixed parameters we can make the `TaggedCache` to have the same methods of the `DefaultCache` + some new methods added. 

We can make so the initial accepred implementation for each `Pcs` to be with the name `\Pcs\NameOfInterface\DefaultNameOfInterface`

Then using the above example for Cache we will have 2 interfaces named `\Pcs\Cache\DefaultCache` and `\Pcs\Cache\TaggedCache` 

Of course this can easily lead of having many interfaces and abuse the standard by having Too-many-standards

I am opening this discussion, so we all can have a thought on this.

**Comment by geggleto** [comment]
Yeah this makes a bit of sense, I honestly don't think we need a bunch of set-in-stone namespaces though. They come naturally through the namespace of a given interface.

`\Pcs\Cache\CacheInterface` ... `\Pcs\Cache\HttpCacheInterface` .. etc

**Comment by Danack** [comment]
I think there are only a couple of things that should be prevented when people create/reserve sub-namespaces:
1. Don't pass off endorsements. e.g. only the people who run the Doctrine project should be allowed to create things that imply endorsement from the Doctrine project. e.g. I should not be allowed to register `Doctrine\Cache` as I am nothing to do with the Doctrine project. Obviously people might need register stuff that does have Doctrine in the name....but it should be more like 'Danack\DoctrineCache` to make it clear the owner is not the doctrine project.
2. There probably should be some rules in place to prevent a 'land-grab' of namespaces - i.e. users must include a specifying name in the namespace to prevent people from grabbing up all the generic useful names.

e.g. I would be allowed to register `DanackCache\CacheInterface` but not `Cache\CacheInterface`
1. I think the top level `\Pcs` namespace could be reserved for things that have `official` approval i.e. I register on the website as `Danack` and them am allowed to start registering things under `Danack`.

Once I've got something that is stable or other people are interested in make a standard to be used across multiple projects, at that point I'm allowed to register it under a `\PCS` namespace.

**Comment by OddGreg** [comment]
_A thought:_

```
<organization>\<classification>\[<subclassifications>\]<entity> 
```

has been useful to me and some others.

ie: `Pcs\Cache\CacheInterface` --> `OddGreg\Cache\DoctrineCache` etc.

Whatever is decided, it should make logical sense and be consistent throughout the groups standards and suggested usage. Me thinks.

**Comment by odan** [comment]
I agree with @OddGreg 
The question is: How can we achieve that the vendor/organization name is unique?
We could suggest that developers use there github username (in PascalCase). 
This essentially allows us to guarantee that each fully qualified class name is unique world wide.

For example:

``` php
namespace Odan\Package\SubPackage;
```
